crates/storage-query-datafusion/src/remote_query_scanner_server.rs

```diff
@@ -0,0 +1,151 @@
+use restate_core::network::{Incoming, MessageRouterBuilder};
+use restate_core::{cancellation_watcher, TaskCenter, TaskKind};
+use restate_types::net::remote_query_scanner::{
+    RemoteQueryScannerClose, RemoteQueryScannerClosed, RemoteQueryScannerNext,
+    RemoteQueryScannerNextResult, RemoteQueryScannerOpen, RemoteQueryScannerOpened, ScannerId,
+};
+use std::collections::BTreeMap;
+use std::pin::Pin;
+use std::sync::Arc;
+use tokio_stream::StreamExt as TokioStreamExt;
+
+type MessageStream<T> = Pin<Box<dyn Stream<Item = Incoming<T>> + Send + Sync + 'static>>;
+
+struct RemoteQueryScannerTask {
+    scanner_id: ScannerId,
+    schema: Schema,
+}
+
+impl RemoteQueryScannerTask {
+    fn new(scanner_id: ScannerId, request: &RemoteQueryScannerOpen) -> Self {
+        let schema: Schema = todo!();
+        Self { scanner_id, schema }
+    }
+
+    async fn next_batch(&self) -> anyhow::Result<Option<Vec<u8>>> {
+        let record_batch: RecordBatch = todo!();
+        let mut buf: Vec<u8> = vec![];
+        {
+            let mut writer = StreamWriter::try_new(&mut buf, &self.schema)?;
+            writer.write(&record_batch)?;
+            writer.finish()?;
+            writer.flush()?;
+        }
+        Ok(Some(buf))
+    }
+
+    async fn run(&self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+impl Drop for RemoteQueryScannerTask {
+    fn drop(&mut self) {
+        // TODO: close the scan object
+        todo!()
+    }
+}
+
+pub struct RemoteQueryScannerServer {
+    task_center: TaskCenter,
+    open_stream: MessageStream<RemoteQueryScannerOpen>,
+    next_stream: MessageStream<RemoteQueryScannerNext>,
+    close_stream: MessageStream<RemoteQueryScannerClose>,
+}
+
+impl RemoteQueryScannerServer {
+    pub fn new(task_center: TaskCenter, router_builder: &mut MessageRouterBuilder) -> Self {
+        let open_stream = router_builder.subscribe_to_stream(64);
+        let next_stream = router_builder.subscribe_to_stream(64);
+        let close_stream = router_builder.subscribe_to_stream(64);
+
+        Self {
+            task_center,
+            open_stream,
+            next_stream,
+            close_stream,
+        }
+    }
+
+    pub async fn run(self) -> anyhow::Result<()> {
+        let RemoteQueryScannerServer {
+            task_center,
+            mut open_stream,
+            mut next_stream,
+            mut close_stream,
+        } = self;
+
+        let mut shutdown = std::pin::pin!(cancellation_watcher());
+        let mut next_scanner_id = 1u64;
+        let mut scanners: BTreeMap<ScannerId, Arc<RemoteQueryScannerTask>> = BTreeMap::default();
+        loop {
+            // Ordered by priority of message types
+            tokio::select! {
+                        biased;
+                        _ = &mut shutdown => {
+                            // stop accepting messages
+                            drop(open_stream);
+                            drop(next_stream);
+                            drop(close_stream);
```

Comment:
I don't believe this is needed since the drop happens implicitly once the variables go out of scope